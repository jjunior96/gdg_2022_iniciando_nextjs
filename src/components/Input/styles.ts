import { lighten } from 'polished';
import styled, { css } from 'styled-components';

import { InputProps } from '.';

type DisableProps = Pick<InputProps, 'isDisabled'>;

type IconContainerProps = {
  isFocused: boolean;
  isErrored?: boolean;
  positionIcon?: 'left' | 'right';
  hiddenInput?: boolean;
} & DisableProps;

export const Container = styled.div`
  width: 100%;
`;

export const IconContainer = styled.div<IconContainerProps>`
  ${({
    theme,
    isErrored,
    isFocused,
    isDisabled,
    positionIcon,
    hiddenInput
  }) => css`
    background: ${theme.colors.white};
    color: ${theme.colors.gray_200};

    border-radius: ${theme.border.radius};
    padding: 0 2rem;

    width: 100%;
    height: 56px;

    font-size: ${theme.font.sizes.medium};

    display: flex;
    align-items: center;

    border: 2px solid ${theme.colors.gray_300};

    cursor: default;

    &:focus-within {
      outline: 3px solid ${lighten(0.4, theme.colors.green)};
    }

    ${positionIcon === 'right' &&
    css`
      > svg {
        order: 2;
      }
    `}

    input {
      color: ${hiddenInput ? 'transparent' : theme.colors.black};
      background: transparent;
      outline: none;
      height: 100%;

      line-height: 1.2;

      font-size: ${theme.font.sizes.xsmall};

      border: 0;

      flex: 1;

      // Remove as setas up e down nos inputs do tipo number
      &::-webkit-outer-spin-button,
      &::-webkit-inner-spin-button {
        -webkit-appearance: none;
        margin: 0;
      }
      &[type='number'] {
        -moz-appearance: textfield;
      }

      /* Remove os estilos nas sugestÃµes do Chrome */
      &:-webkit-autofill,
      &:-webkit-autofill:hover,
      &:-webkit-autofill:focus {
        -webkit-text-fill-color: ${theme.colors.gray_300};
        -webkit-box-shadow: #282f3a26 inset;
        box-shadow: #282f3a26;
        transition: background-color 5000s ease-in-out 0s;
      }

      &::placeholder {
        color: ${theme.colors.gray_200};
      }

      ${isDisabled &&
      css`
        cursor: not-allowed;

        color: ${theme.colors.gray_200};
      `}
    }

    svg {
      margin-right: ${theme.spacings.xsmall};
      color: ${theme.colors.gray_300};
    }

    ${isFocused &&
    css`
      border-color: ${theme.colors.green};
      color: ${theme.colors.green};

      > svg {
        color: ${theme.colors.green};
      }
    `}

    ${isErrored &&
    css`
      border: 2px solid ${theme.colors.error};

      > svg {
        color: ${theme.colors.error};
      }
    `}




    ${isDisabled &&
    css`
      cursor: not-allowed;

      border-color: ${lighten(0.3, theme.colors.gray_200)};

      color: ${lighten(0.3, theme.colors.gray_300)};

      > svg {
        color: ${lighten(0.3, theme.colors.gray_300)};
      }
    `}

    &:hover {
      border-color: ${theme.colors.black};
    }
  `}
`;

export const ButtonVisiblePassword = styled.button`
  ${({ theme }) => css`
    text-align: center;

    border: none;
    background: none;

    border-radius: 10%;

    width: 2.4rem;
    height: 2.4rem;

    font-size: 0.6rem;

    &:focus-within {
      outline: 3px solid ${lighten(0.2, theme.colors.green)};

      > svg {
        color: ${theme.colors.green};
      }
    }

    svg {
      width: 100%;
      height: 100%;
    }
  `}
`;
