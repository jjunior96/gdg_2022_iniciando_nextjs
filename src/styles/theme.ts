export default {
  grid: {
    container: '96rem',
    gutter: '3.2rem'
  },
  components: {
    sidebar: '20rem',
    header: '8rem'
  },
  border: {
    radius: '0.4rem'
  },
  box: {
    shadow: '0 8px 8px rgba(0, 0, 0, 0.2)'
  },
  font: {
    family:
      "Montserrat, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif",
    light: 300,
    normal: 400,
    bold: 600,
    sizes: {
      xxxxsmall: '0.8rem',
      xxxsmall: '1.2rem',
      xxsmall: '1.4rem',
      xsmall: '1.6rem',
      small: '2rem',
      medium: '2.4rem',
      large: '3.2rem',
      xlarge: '4rem',
      xxlarge: '4.8rem',
      xxxlarge: '5.6rem',
      huge: '6.4rem'
    }
  },
  colors: {
    primary: '#24A66B',
    secundary: '#5855E9',
    white: '#FAFAFA',
    gray_50: '#f0efeb',
    gray_100: '#D9D9D9',
    gray_200: '#91AAB4',
    gray_300: '#3E606F',
    gray_450: '#193441',
    gray_500: '#343a40',
    yellow_300: '#F1C40F',
    purple_100: '#7F57DF',
    purple_300: '#5855E9',
    green: '#24A66B',
    red: '#E63737',
    success: '#24A66B',
    error: '#E63737',
    info: '#189CEF',
    black: '#0D0D0D'
  },
  spacings: {
    xxxsmall: '0.8rem',
    xxsmall: '1rem',
    xsmall: '1.6rem',
    small: '2.4rem',
    medium: '3.2rem',
    large: '4.0rem',
    xlarge: '4.8rem',
    xxlarge: '5.6rem'
  },
  layers: {
    base: 10,
    menu: 20,
    overlay: 30,
    modal: 40,
    alwaysOnTop: 50
  },
  modal: {
    sizes: {
      small: '40rem',
      medium: '48rem',
      large: '54rem'
    }
  },
  transition: {
    default: '0.3s ease-in-out',
    fast: '0.1s ease-in-out'
  }
} as const;
