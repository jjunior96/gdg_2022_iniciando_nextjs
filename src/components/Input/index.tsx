import { InputHTMLAttributes, useState, useCallback, forwardRef } from 'react';
import {
  AiOutlineEyeInvisible as NotVisibleIcon,
  AiOutlineEye as VisibleIcon
} from 'react-icons/ai';

import ErrorText from 'components/ErrorText';
import LabelInput, { LabelColorType } from 'components/LabelInput';

import * as S from './styles';

type Props = {
  label?: string;
  labelColor?: LabelColorType;
  isRequired?: boolean;
  error?: string;
  isDisabled?: boolean;
  positionIcon?: 'left' | 'right';
  hiddenInput?: boolean;
  icon?: React.ReactNode;
};

export type InputProps = InputHTMLAttributes<HTMLInputElement> & Props;

const Input: React.ForwardRefRenderFunction<HTMLInputElement, InputProps> = (
  {
    label,
    labelColor = 'black',
    name,
    isRequired = false,
    error,
    isDisabled = false,
    type = 'text',
    positionIcon = 'left',
    hiddenInput = false,
    icon: Icon,
    ...rest
  },
  ref
) => {
  const [isFocused, setIsFocused] = useState(false);
  const [isVisiblePassword, setIsVisiblePassword] = useState(false);

  const typeVisible = isVisiblePassword ? 'text' : 'password';
  const typeChangePassword = type === 'password' ? typeVisible : type;

  const handleInputFocus = useCallback(() => {
    setIsFocused(true);
  }, []);

  const handleInputBlur = useCallback(() => {
    setIsFocused(false);
  }, []);

  const handleVisiblePassword = useCallback(() => {
    setIsVisiblePassword((oldState) => !oldState);
  }, []);

  return (
    <S.Container>
      <LabelInput
        htmlFor={label}
        isRequired={isRequired}
        isDisabled={isDisabled}
        labelColor={labelColor}
      >
        {label}
      </LabelInput>

      <S.IconContainer
        isFocused={isFocused}
        isErrored={!!error}
        isDisabled={isDisabled}
        positionIcon={positionIcon}
        hiddenInput={hiddenInput}
        data-testid="icon-container"
      >
        {Icon}
        <input
          {...rest}
          data-testid="input-element"
          name={name}
          type={typeChangePassword}
          onFocus={handleInputFocus}
          onBlur={handleInputBlur}
          disabled={isDisabled}
          ref={ref}
        />

        {type === 'password' && (
          <S.ButtonVisiblePassword
            type="button"
            onClick={handleVisiblePassword}
            aria-label="button-visible-password"
          >
            {isVisiblePassword ? (
              <VisibleIcon aria-label="is-visible" />
            ) : (
              <NotVisibleIcon aria-label="not-visible" />
            )}
          </S.ButtonVisiblePassword>
        )}
      </S.IconContainer>

      {error && <ErrorText>{error}</ErrorText>}
    </S.Container>
  );
};

export default forwardRef(Input);
