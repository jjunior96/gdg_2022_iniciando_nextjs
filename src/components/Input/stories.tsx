import { BiSearch as SearchIcon } from 'react-icons/bi';
import { BsPerson as PersonIcon } from 'react-icons/bs';

import { Story, Meta } from '@storybook/react/types-6-0';

import Input, { InputProps } from '.';

export default {
  title: 'Form/Input',
  component: Input,
  argTypes: {
    icon: {
      type: ''
    }
  }
} as unknown as Meta;

export const Default: Story<InputProps> = (args) => (
  <div
    style={{
      width: '40rem',
      margin: '0 auto',
      paddingTop: '1rem'
    }}
  >
    <Input {...args} />
  </div>
);

export const withLabel: Story<InputProps> = (args) => (
  <div
    style={{
      width: '40rem',
      margin: '0 auto'
    }}
  >
    <Input {...args} />
  </div>
);

withLabel.args = {
  label: 'Name',
  placeholder: 'Insert user name'
};

export const withError: Story<InputProps> = (args) => (
  <div
    style={{
      width: '40rem',
      margin: '0 auto'
    }}
  >
    <Input error="Field is required" isRequired {...args} />
  </div>
);

withError.args = {
  label: 'Name',
  placeholder: 'Insert user name',
  icon: <PersonIcon />
};

export const withIcon: Story<InputProps> = (args) => (
  <div
    style={{
      width: '40rem',
      margin: '0 auto'
    }}
  >
    <Input {...args} />
  </div>
);

withIcon.args = {
  icon: <SearchIcon />,
  label: 'Name',
  placeholder: 'Search here'
};

export const disable: Story<InputProps> = (args) => (
  <div
    style={{
      width: '40rem',
      margin: '0 auto'
    }}
  >
    <Input label="Search" isDisabled {...args} />
  </div>
);

disable.args = {
  icon: <SearchIcon />,
  placeholder: 'Search here'
};
