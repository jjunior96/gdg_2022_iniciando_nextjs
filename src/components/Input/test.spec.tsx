import { fireEvent, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { renderWithTheme } from 'utils/tests/__config__/helpers';

import theme from 'styles/theme';

import Input from '.';

describe('<Input />', () => {
  it('should render the Input component', () => {
    const { container } = renderWithTheme(
      <Input name="Name" aria-label="name-input" />
    );

    expect(
      screen.getByRole('textbox', { name: /name-input/i })
    ).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the Input component with label', () => {
    renderWithTheme(<Input label="Name" />);

    expect(screen.getByText(/name/i)).toHaveStyle({
      'font-size': theme.font.sizes.xxsmall,
      color: theme.colors.black
    });
  });

  it('should render the Input component with labelColor', () => {
    renderWithTheme(<Input label="Name" labelColor="white" />);

    expect(screen.getByText(/name/i)).toHaveStyle({
      color: theme.colors.white
    });
  });

  it('should render the Input component when is required', () => {
    renderWithTheme(<Input label="Name" isRequired />);

    // Quando `isRequired` é passado, é exibido um span com "*" para
    // simbolizar que o campo é obrigatório
    expect(screen.getByText(/\*/)).toHaveStyle({
      display: 'inline',
      color: theme.colors.error
    });
  });

  it('should render the Input component when error is passed', () => {
    renderWithTheme(<Input label="Name" error="Insert a name valid" />);

    expect(screen.getByText(/Insert a name valid/i)).toBeInTheDocument();
  });

  it('should to show icon (visible or notVisible) when input type password', () => {
    renderWithTheme(
      <Input label="Password" name="password-user" type="password" />
    );

    const inputElement = screen.getByTestId('input-element');

    // Botao de alterar visibilidade da senha
    const buttonSetVisiblePassword = screen.getByLabelText(
      /button-visible-password/
    );

    // Deve exibir o icon "NotIsVisible"
    expect(buttonSetVisiblePassword).toBeInTheDocument();

    const iconNotVisible = screen.getByLabelText(/not-visible/);

    expect(iconNotVisible).toBeInTheDocument();

    // O type do input deve ser 'password' por default
    expect(inputElement).toHaveAttribute('type', 'password');

    // Clique no botao para visualizar a senha
    userEvent.click(buttonSetVisiblePassword);

    const iconVisible = screen.getByLabelText(/is-visible/);

    // Deve exibir o icon "IsVisible"
    expect(iconVisible).toBeInTheDocument();
    expect(iconNotVisible).not.toBeInTheDocument();

    // O type do input deve ser 'text'
    expect(inputElement).toHaveAttribute('type', 'text');
  });

  it('should is not accessible by tab when disabled', () => {
    renderWithTheme(<Input label="Name" isDisabled />);

    const input = screen.getByRole('textbox');
    expect(document.body).toHaveFocus();

    userEvent.tab();

    expect(input).not.toHaveFocus();
  });

  it('should change border color input when hover', async () => {
    renderWithTheme(<Input label="Name" isDisabled />);

    const input = screen.getByTestId('icon-container');

    userEvent.hover(input);

    waitFor(() => {
      expect(input).toHaveStyle({
        border: `2px solid ${theme.colors.black}`
      });
    });

    userEvent.unhover(input);

    expect(input).toHaveStyle({
      border: `2px solid ${theme.colors.gray_300}`
    });
  });

  it('should change focus and blur input', async () => {
    renderWithTheme(<Input label="Name" isDisabled />);

    const input = screen.getByRole('textbox');

    fireEvent.focus(input);

    const iconContainer = screen.getByTestId('icon-container');

    waitFor(() => {
      expect(iconContainer).toHaveStyle({
        border: `2px solid ${theme.colors.green}`
      });
      expect(iconContainer).toHaveFocus();
    });

    fireEvent.blur(input);

    expect(iconContainer).toHaveStyle({
      border: `2px solid ${theme.colors.gray_300}`
    });
    expect(iconContainer).not.toHaveFocus();
  });
});
