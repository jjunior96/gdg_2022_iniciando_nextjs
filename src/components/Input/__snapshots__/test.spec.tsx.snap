// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should render the Input component 1`] = `
.c1 {
  font-size: 1.4rem;
  margin: 0;
  padding: 0;
  line-height: 1;
  color: #0D0D0D;
}

.c1 span {
  display: none;
}

.c0 {
  width: 100%;
}

.c2 {
  background: #FAFAFA;
  color: #91AAB4;
  border-radius: 0.4rem;
  padding: 0 2rem;
  width: 100%;
  height: 56px;
  font-size: 2.4rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 2px solid #3E606F;
  cursor: default;
}

.c2:focus-within {
  outline: 3px solid #aaecce;
}

.c2 input {
  color: #0D0D0D;
  background: transparent;
  outline: none;
  height: 100%;
  line-height: 1.2;
  font-size: 1.6rem;
  border: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c2 input::-webkit-outer-spin-button,
.c2 input::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.c2 input[type='number'] {
  -moz-appearance: textfield;
}

.c2 input:-webkit-autofill,
.c2 input:-webkit-autofill:hover,
.c2 input:-webkit-autofill:focus {
  -webkit-text-fill-color: #3E606F;
  -webkit-box-shadow: #282f3a26 inset;
  box-shadow: #282f3a26;
  -webkit-transition: background-color 5000s ease-in-out 0s;
  transition: background-color 5000s ease-in-out 0s;
}

.c2 input::-webkit-input-placeholder {
  color: #91AAB4;
}

.c2 input::-moz-placeholder {
  color: #91AAB4;
}

.c2 input:-ms-input-placeholder {
  color: #91AAB4;
}

.c2 input::placeholder {
  color: #91AAB4;
}

.c2 svg {
  margin-right: 1.6rem;
  color: #3E606F;
}

.c2:hover {
  border-color: #0D0D0D;
}

<div
  class="c0"
>
  <label
    class="c1"
    for=""
  >
     
    <span>
      *
    </span>
  </label>
  <div
    class="c2"
    data-testid="icon-container"
  >
    <input
      aria-label="name-input"
      data-testid="input-element"
      name="Name"
      type="text"
    />
  </div>
</div>
`;
